cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -820,6 +844,30 @@ protected Future<Boolean> delete(Reconciliation reconciliation) {
                 .map(Boolean.FALSE); // Return FALSE since other resources are still deleted by garbage collection
     }
 
+    /**
+     * Remove the metrics specific to the kind implementing it.
+     *
+     * @param desiredNames  Set of resources which should be reconciled
+     * @param namespace     The namespace to reconcile, or {@code *} to reconcile across all namespaces.
+     */
+    @Override
+    public void removeMetrics(Set<NamespaceAndName> desiredNames, String namespace) {
+        // Removing all CA cluster metrics not in the desired set
+        Set<String> desiredClusterNames = desiredNames.stream().map(NamespaceAndName::getName).collect(Collectors.toSet());
+        metrics.removeMetricsForCertificates(key -> {
+            if (!key.getKind().equals(kind())) {
+                return false;
+            }
+            if (namespace.equals(ConfigParameter.ANY_NAMESPACE)) {
+                Set<String> desiredNamespaces = desiredNames.stream().map(NamespaceAndName::getNamespace).collect(Collectors.toSet());
+                return !desiredNamespaces.contains(key.getNamespace()) && !desiredClusterNames.contains(key.getClusterName());
+            }
+
+            // Otherwise only remove the ones related to the namespace and not in the desired set
+            return key.getNamespace().equals(namespace) && !desiredClusterNames.contains(key.getClusterName());
+        });
+    }
+
```

Comment:
I hate to be PITA. But have you actually tried what I suggested before:
```
    public void removeMetrics(Set<NamespaceAndName> desiredNames, String namespace) {
        metrics.removeMetricsForCertificates(key -> {
            return !desiredNames.contains(new NamespaceAndName(key.getNamespace(), key.getClusterName()));
        });
    }
```

This code bothers me because - while I think it should work in general - the parts about comparing kind and about the `ANY_NAMESPACE` shoud IMHO be never triggered. If we have in the KafkaAsemblyOperator any metrics for a different kind, that indicates some bug. The same for if we have any metrics for the `*` namespace.

So I think we should:
* Remove them if they are a dead code that is never triggered
* Investigate them deeper if they are ever triggered.

Reply:
I also guess we still don't have any unit test for this.

Reply:
Sure we can try this, but I don't see you using the `namespace` parameter, but only from the desired set, is that a mistake or by intention?

Reply:
Alright, it differs from my notes from our meeting, but that's completely fine, let's make sure we get the desired outcome of the PR.

Let me do the changes you suggested and remove the `namespace` parameter and do a thorough test to verify.

Reply:
Sorry, I guess I missed this comment yesterday. Yeah, the idea was to have a unit test for `KafkaAssemblyOperator.removeMetrics` ... or more specifically I guess the lambda that is inside that method. Hence the idea that if you move the lambda into a static method, it would be easier to test it (as testing the whole `KafkaAssemblyOperator` chain can be quite ocmplicated)